hoomd_benchmarks/hpmc_pair_union_wca.py

```diff
@@ -0,0 +1,145 @@
+# Copyright (c) 2021-2023 The Regents of the University of Michigan
+# Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+"""Lennard-Jones HPMC pair potential benchmark."""
+
+import hoomd
+import numpy
+
+from . import common, hpmc_base
+from .configuration.hard_sphere import make_hard_sphere_configuration
+
+DEFAULT_MODE = 'compiled'
+DEFAULT_LEAF_CAPACITY = 0
+DEFAULT_GRID = 4
+
+
+class HPMCPairUnionWCA(hpmc_base.HPMCBenchmark):
+    """Test performance of HPMC union potentials."""
+
+    def __init__(
+        self,
+        mode=DEFAULT_MODE,
+        leaf_capacity=DEFAULT_LEAF_CAPACITY,
+        grid=DEFAULT_GRID,
+        **kwargs,
+    ):
+        self.mode = mode
+        self.leaf_capacity = leaf_capacity
+        self.grid = grid
+        super().__init__(**kwargs)
+
+    @staticmethod
+    def make_argument_parser():
+        """Make an ArgumentParser instance for benchmark options."""
+        parser = common.Benchmark.make_argument_parser()
+        parser.add_argument('--mode', default=DEFAULT_MODE, help='Compute mode.')
+        parser.add_argument(
+            '--leaf-capacity',
+            default=DEFAULT_LEAF_CAPACITY,
+            help='Leaf capacity.',
+            type=int,
+        )
+        parser.add_argument(
+            '--grid',
+            default=DEFAULT_GRID,
+            help='Number of grid points along an edge.',
+            type=int,
+        )
+        return parser
+
+    @classmethod
+    def runs_on_device(cls, device):
+        """Returns True when the benchmark can be run on the given device."""
+        if isinstance(device, hoomd.device.GPU):
+            return False
+
+        pair_class = getattr(hoomd.hpmc.pair, 'Union', None)
+        if pair_class is None:
+            # Skip this benchmark on HOOMD releases that lack the needed pair
+            # potential.
+            return False
+
+        return True
+
+    def make_simulation(self):
+        """Make the Simulation object."""
+        path = make_hard_sphere_configuration(
+            N=self.N,
+            rho=self.rho,
+            dimensions=self.dimensions,
+            device=self.device,
+            verbose=self.verbose,
+        )
+
+        integrator = hoomd.hpmc.integrate.Sphere(default_d=0.3, default_a=0.4)
+        integrator.shape['A'] = dict(diameter=0, orientable=True)
+
+        sim = hoomd.Simulation(device=self.device, seed=10)
+        sim.create_state_from_gsd(filename=str(path))
+
+        sigma = 0.1
+        r_cut = 2 ** (1 / 6) * sigma
+        points = numpy.linspace(start=-0.5, stop=0.5, num=self.grid)
+        positions = []
+        positions.extend([(x, 0, 0) for x in points])
+        positions.extend([(0, x, 0) for x in points])
+        positions.extend([(0, 0, x) for x in points])
```

Comment:
Wouldn't this result in overlaps if grid is chosen poorly? Also. this doesn't form a grid as points are only placed on the axes lines. That isn't necessarily a problem, just makes the argument name grid seem odd.

Reply:
Yes, it should be named "spacing" or something like that. This benchmark attempts to prevent overlaps by keeping the overall particles the same size and placing overlapping constituents.

It is not a very realistic placement, but I used it to get some idea of what union size was the minimum needed for `leaf_capacity=4` to be faster than `leaf_capacity=0`.